Snakefile

```diff
@@ -0,0 +1,402 @@
+import subprocess
+
+#################################################################
+####################### general setup ###########################
+#################################################################
+
+# certain files can be fetched from the database or downloaded from our S3 bucket.
+# config["fetch_from_database"] dictates this
+# We could implement this via remote snakemake inputs (such as in nextstrain/ncov),
+# conditional rules (a bit of an anti-pattern) or ruleorders.
+ruleorder: download_previous_biosample > fetch_biosample_from_database
+ruleorder: download_previous_raw_ndjson > fetch_from_database
+if config.get("fetch_from_database", False):
+    ruleorder: fetch_biosample_from_database > download_previous_biosample
+    ruleorder: fetch_from_database > download_previous_raw_ndjson
```

Comment:
Hmm. I agree that special input functions as conditionals are best avoided if possible, but not sure I find ruleorders much better. We could instead conditionalize the rule _definitions_ themselves (either at the whole `rule` level or at the `shell/run` block level) with normal Python `if/elif/else` or standard function calls instead of using ruleorder to mask some rules. Not a blocking concern, just a thought.

Reply:
This seems like a case where asking ourselves how we would implement this logic in a different workflow language might lead to a clearer solution. In a less expressive workflow language, we might be forced to move this kind of conditional logic into a single script that takes `--fetch-from-database` as an argument and produces the output(s) supported by either downloading from S3 or GISAID, for example.

This makes the workflow simpler at the expense of more complicated scripts.

Reply:
Switched to python conditionals within the rules, which I agree is nicer. Commit ab59826